![header](https://apifriends.com/wp-content/uploads/2018/08/hacker-news.jpg)

# Dataset 

This data set is Hacker News posts **from the last 12 months of the year 2016**.It includes the following columns:

- title: title of the post (self explanatory)

- url: the url of the item being linked to

- num_points: the number of upvotes the post received

- num_comments: the number of comments the post received

- author: the name of the account that made the post

- created_at: the date and time the post was made (the time zone is Eastern Time in the US)


# Context

Hacker News is a popular technology site that was started by Y Combinator. Users submit stories (a post) and vote and/comment upon. Most popular posts arrive at the top and can get over 100k visitors as a result. Publications are under strict [guidelines](https://news.ycombinator.com/newsguidelines.html) and it is not easy to stand out. 

We will try here to determine at what time posts should be posted in the hopes of getting the more *Up Votes* and *Comments* and become popular. 

# Data Cleaning

In [15]:
from csv import reader #this package will allow us to read and manipulate the csv file

opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)

print(hn[:5]) #we print the first five rows to have quick understanding of the dataset composition

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [16]:
headers = hn[0] #we isolate the header in order to remove it from the rest of the dataset
print(headers)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [17]:
hn = hn[1:] #the header is removed 
print(hn[:5])

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


We opened the dataset using the *reader* package from **csv**. After printing the 5 first rows of the data set (named **hn**). We can see that the configuration corresponds to the [information given above](#Dataset). 

# Number of posts by category

In [18]:
ask_posts=[]
show_posts=[]
other_posts=[]

for row in hn: 
    title = row[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(row)
    elif title.lower().startswith("show hn"):
        show_posts.append(row)
    else: 
        other_posts.append(row)
        
print("Number of posts in Ask posts:", len(ask_posts))
print("Number of posts in Show posts:", len(show_posts))
print("Number of posts of other post types:", len(other_posts))

Number of posts in Ask posts: 1744
Number of posts in Show posts: 1162
Number of posts of other post types: 17194


Here, we separated the different kind of posts to make it easier to analyze. In the end, there are more people asking questions than posting other topics. **Ask posts** represents 8,67% of the total of posts on *Hacker News*. 

# Finding out the number of comments by the hour

In [19]:
total_ask_comments = 0

for row in ask_posts:
    total_ask_comments += int(row[4])

avg_ask_comments = total_ask_comments / len(ask_posts)
print("Average number of comments in Ask posts:", avg_ask_comments)

Average number of comments in Ask posts: 14.038417431192661


In [20]:
total_show_comments = 0

for row in show_posts: 
    total_show_comments += int(row[4])

avg_show_comments = total_show_comments / len(show_posts)
print("Average number of comments in Show posts:", avg_show_comments)

Average number of comments in Show posts: 10.31669535283993


In [21]:
import datetime as dt 

result_list = []

for post in ask_posts: 
    result_list.append(
        [post[6], int(post[4])]
    )
    
count_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for row in result_list: 
    hours = row[0]
    comment = row[1]
    time = dt.datetime.strptime(hours, date_format).strftime("%H")
    if time not in count_by_hour: 
        comments_by_hour[time] = comment
        count_by_hour[time] = 1
    else: 
        comments_by_hour[time] += comment
        count_by_hour[time] += 1

comments_by_hour

{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

It seems that **3pm** is the time where people comment the most, the total sum of comments at this hour is **4477**. Let's dive deeper to see if we are correct.

In [22]:
avg_by_hour = []

for comment in comments_by_hour:
    avg_by_hour.append([comment, comments_by_hour[comment] / count_by_hour[comment]])
    
avg_by_hour

[['17', 11.46],
 ['07', 7.852941176470588],
 ['23', 7.985294117647059],
 ['13', 14.741176470588234],
 ['20', 21.525],
 ['08', 10.25],
 ['10', 13.440677966101696],
 ['19', 10.8],
 ['09', 5.5777777777777775],
 ['12', 9.41095890410959],
 ['18', 13.20183486238532],
 ['04', 7.170212765957447],
 ['16', 16.796296296296298],
 ['21', 16.009174311926607],
 ['00', 8.127272727272727],
 ['11', 11.051724137931034],
 ['03', 7.796296296296297],
 ['15', 38.5948275862069],
 ['05', 10.08695652173913],
 ['01', 11.383333333333333],
 ['14', 13.233644859813085],
 ['22', 6.746478873239437],
 ['06', 9.022727272727273],
 ['02', 23.810344827586206]]

In [23]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse=True)

print("Top 5 Hours for Ask Posts Comments")
for avg, hour in sorted_swap: 
    print(
        "{}: {:.2f} average comments per post".format(dt.datetime.strptime(hour, "%H").strftime("%H:%M"),avg))
    

[[11.46, '17'], [7.852941176470588, '07'], [7.985294117647059, '23'], [14.741176470588234, '13'], [21.525, '20'], [10.25, '08'], [13.440677966101696, '10'], [10.8, '19'], [5.5777777777777775, '09'], [9.41095890410959, '12'], [13.20183486238532, '18'], [7.170212765957447, '04'], [16.796296296296298, '16'], [16.009174311926607, '21'], [8.127272727272727, '00'], [11.051724137931034, '11'], [7.796296296296297, '03'], [38.5948275862069, '15'], [10.08695652173913, '05'], [11.383333333333333, '01'], [13.233644859813085, '14'], [6.746478873239437, '22'], [9.022727272727273, '06'], [23.810344827586206, '02']]
Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post
13:00: 14.74 average comments per post
10:00: 13.44 average comments per post
14:00: 13.23 average comments per post
18:00: 13.20 average comments per post
17:00:

The time slot that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set documentation, the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.